### PR TITLE
Ci 342 and ci 344 randomise project pages on project page

### PR DESCRIPTION
**Pull request by jessgilbert**
To do (delete all that do not apply):

- [x] Change has a jira ticket that has the correct status.
- [x] Changelog entry added.

### CODE REVIEW

File: core/serializers.py [hunk 1]
```diff
@@ -77,3 +79,32 @@ def get_parent_page_data_for(self, instance, serializer):
...
+        page_type = CapitalInvestOpportunityPage
+        all_opportunity_pages = page_type.objects.live().public()
+
+        related_sectors_with_opps = {
+            sector['title']: [] for sectors in related_sectors
+            for sector in sectors.values() if sector
+        }
+
+        for opportunity in all_opportunity_pages:
+            for sector in related_sectors_with_opps.keys():
```
[the remark this answers is not in the document]
Comment (jessgilbert) [reply core/serializers.py]: Oh awesome I didn't know that, thanks have done it :)

File: core/serializers.py [hunk 2]
```diff
@@ -77,3 +79,32 @@ def get_parent_page_data_for(self, instance, serializer):
...
+        all_opportunity_pages = page_type.objects.live().public()
+
+        related_sectors_with_opps = {
+            sector['title']: [] for sectors in related_sectors
+            for sector in sectors.values() if sector
+        }
+
+        for opportunity in all_opportunity_pages:
+            for sector in related_sectors_with_opps:
+                for related_sector in opportunity.related_sectors.all():
```
Comment (sdonk) [inline core/serializers.py]: This is going to hit the database on every cycle (callable attributes are not cached in Django). Can you please assign this to a variable so it's cached

Comment (jessgilbert) [reply core/serializers.py]: Sure, as in:
```
            for sector in related_sectors_with_opps:
                opps_related_sectors = opportunity.related_sectors.all()
                for related_sector in opps_related_sectors:
``` ?

File: core/serializers.py [hunk 3]
```diff
@@ -77,3 +79,33 @@ def get_parent_page_data_for(self, instance, serializer):
...
         )
         return serializer.data
+
+
+class SameSectorOpportunitiesHelper(serializers.Serializer):
+    def get_same_sector_opportunity_pages_data_for(
+            self, instance, serializer, related_sectors
+    ):
+        page_type = CapitalInvestOpportunityPage
+        all_opportunity_pages = page_type.objects.live().public()
```
Comment (sdonk) [inline core/serializers.py]: prefetc_related('related_sectors') should save you some queries

File: core/serializers.py (hunk 2 above)
Comment (sdonk) [reply core/serializers.py]: yep

File: core/serializers.py (hunk 3 above)
Comment (jessgilbert) [reply core/serializers.py]: Awesome! Didn't know about that either, done 😀

**Pull request merged**